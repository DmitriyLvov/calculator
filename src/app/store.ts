import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";

import { configureStore } from "@reduxjs/toolkit";

import calculatorReducer from "../slices/CalculatorSlice";

export const store = configureStore({
  reducer: {
    calculator: calculatorReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
