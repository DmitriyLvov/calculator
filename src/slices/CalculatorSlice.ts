import { CalcComponentType } from 'components/Views/ConstructorPanel/ConstructorPanel';
import { ISign } from 'components/Views/ConstructorPanel/SignPanel/SignPanel';

import { createSlice } from '@reduxjs/toolkit';

interface IInitialState {
  mode: "Constructor" | "Runtime";
  operation?: ISign;
  isDecimal: boolean;
  firstNumber: number;
  displayNumber: number;
  displayText: string;
  specification: CalcComponentType[];
}

const initialState: IInitialState = {
  mode: "Constructor",
  specification: [],
  displayNumber: 0,
  firstNumber: 0,
  displayText: "0",
  isDecimal: false,
};

export const counterSlice = createSlice({
  name: "counter",
  initialState,
  reducers: {
    setDisplay: (state, action) => {
      const { textFormat, numberFormat } = action.payload;
      state.displayText = textFormat;
      state.displayNumber = numberFormat;
    },
    setOperation: (state, action) => {
      state.operation = action.payload;
    },
    setDecimal: (state, action) => {
      state.isDecimal = action.payload;
    },
    saveNumber: (state) => {
      state.firstNumber = state.displayNumber;
      state.displayNumber = 0;
    },
    setMode: (state, action) => {
      if (state.mode !== action.payload) {
        state.mode = action.payload;
        state.displayText = "0";
        state.displayNumber = 0;
      }
    },
    changeSpecificaion: (state, action) => {
      const { operation } = action.payload;
      if (operation === "addElement") {
        const { type, index, currentIndex } = action.payload;
        if (currentIndex === -1) {
          state.specification.splice(index, 0, type);
        } else {
          state.specification.splice(
            index,
            0,
            state.specification.splice(currentIndex, 1)[0]
          );
        }
      }
      if (operation === "deleteElement") {
        const { index } = action.payload;
        state.specification.splice(index, 1);
      }
    },
  },
});

export const {
  setDisplay,
  saveNumber,
  setMode,
  changeSpecificaion,
  setOperation,
  setDecimal,
} = counterSlice.actions;

export default counterSlice.reducer;
