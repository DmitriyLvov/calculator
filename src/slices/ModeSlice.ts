import { CalcComponentType } from "components/Views/ConstructorPanel/ConstructorPanel";

import { createSlice } from "@reduxjs/toolkit";

type IInitialState = {
  mode: "Constructor" | "Runtime";
  specification: CalcComponentType[];
};

const initialState: IInitialState = { mode: "Constructor", specification: [] };

export const modeSlice = createSlice({
  name: "mode",
  initialState,
  reducers: {
    setMode: (state, action) => {
      if (state.mode !== action.payload) {
        state.mode = action.payload;
      }
    },
    changeSpecificaion: (state, action) => {
      const { operation } = action.payload;
      if (operation === "addElement") {
        const { type, index, currentIndex } = action.payload;
        if (currentIndex === -1) {
          state.specification.splice(index, 0, type);
        } else {
          const correctedIndex = currentIndex < index ? index - 1 : index;
          state.specification.splice(
            correctedIndex,
            0,
            state.specification.splice(currentIndex, 1)[0]
          );
        }
      }
      if (operation === "deleteElement") {
        const { index } = action.payload;
        state.specification.splice(index, 1);
      }
    },
  },
});

export const { changeSpecificaion, setMode } = modeSlice.actions;

export default modeSlice.reducer;
