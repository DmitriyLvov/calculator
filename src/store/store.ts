import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";

import {
  combineReducers,
  configureStore,
  PreloadedState,
} from "@reduxjs/toolkit";

import calculatorReducer from "../slices/CalculatorSlice";
import modeReducer from "../slices/ModeSlice";

const rootReducer = combineReducers({
  calculator: calculatorReducer,
  mode: modeReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export const store = setupStore();

export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
