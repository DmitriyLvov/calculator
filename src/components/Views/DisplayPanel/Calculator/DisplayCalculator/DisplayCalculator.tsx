import DigitDisplay from "components/business/DigitDisplay/DigitDisplay";
import EqualButton from "components/business/EqualButton/EqualButton";
import { CalcComponentType } from "components/Views/ConstructorPanel/ConstructorPanel";
import NumberPanel from "components/Views/ConstructorPanel/NumberPanel/NumberPanel";
import SignPanel from "components/Views/ConstructorPanel/SignPanel/SignPanel";
import useHotKeys from "hooks/useHotKeys";
import React, { useEffect } from "react";
import { useAppSelector } from "store/store";

import styles from "../ConstructorCalculator/ConstructorCalculator.module.scss";
import DragElement from "../ConstructorCalculator/DragElement/DragElement";

type IModule = {
  [key in CalcComponentType]: React.ReactElement;
};

const modules: IModule = {
  Display: <DigitDisplay />,
  Numbers: <NumberPanel />,
  Signs: <SignPanel />,
  Equal: <EqualButton />,
};

const DisplayCalculator = () => {
  const { specification } = useAppSelector((state) => state.calculator);

  const keyExecute = useHotKeys();

  const keyHandler = (e: KeyboardEvent) => {
    e.preventDefault();
    keyExecute(e.key);
  };

  useEffect(() => {
    document.addEventListener("keydown", keyHandler);
    return () => document.removeEventListener("keydown", keyHandler);
  }, []);

  return (
    <div className={styles.component}>
      {specification.map((component, index) => (
        <DragElement key={index} index={index}>
          {modules[component]}
        </DragElement>
      ))}
      <DragElement
        key={specification.length}
        index={specification.length}
        isLast={true}
      />
    </div>
  );
};

export default DisplayCalculator;
