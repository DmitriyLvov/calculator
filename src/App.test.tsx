import { addDigit, resetData, setOperation } from "slices/CalculatorSlice";
import { RootState, setupStore } from "store/store";
import { renderWithProviders } from "utils/test/renderWithProviders";

import { fireEvent, screen } from "@testing-library/react";

import DisplayCalculator from "./components/Views/DisplayPanel/Calculator/DisplayCalculator/DisplayCalculator";

const preloadedState: RootState = {
  mode: {
    specification: ["Numbers", "Equal", "Signs", "Display"],
    mode: "Runtime",
  },
  calculator: {
    firstNumber: "0",
    resultWasShown: false,
    displayNumber: "0",
  },
};

const store = setupStore(preloadedState);

describe("Test Render", () => {
  it("Snapshot", () => {
    const component = renderWithProviders(<DisplayCalculator />, { store });
    expect(component).toMatchSnapshot();
  });
});

describe("Test Calculations", () => {
  it("Add operation", () => {
    store.dispatch(addDigit("0,1"));
    store.dispatch(setOperation("+"));
    store.dispatch(addDigit("0,2"));
    renderWithProviders(<DisplayCalculator />, { store });
    fireEvent.click(screen.getByRole("button", { name: "=" }));
    expect(screen.getByText("0,3")).toBeInTheDocument();
  });
  it("Minus operation", () => {
    store.dispatch(resetData());
    store.dispatch(addDigit("90"));
    store.dispatch(setOperation("-"));
    store.dispatch(addDigit("70"));
    renderWithProviders(<DisplayCalculator />, { store });
    fireEvent.click(screen.getByRole("button", { name: "=" }));
    expect(screen.getByText("20")).toBeInTheDocument();
  });
  it("Multiply operation", () => {
    store.dispatch(resetData());
    store.dispatch(addDigit("9"));
    store.dispatch(setOperation("x"));
    store.dispatch(addDigit("7"));
    renderWithProviders(<DisplayCalculator />, { store });
    fireEvent.click(screen.getByRole("button", { name: "=" }));
    expect(screen.getByText("63")).toBeInTheDocument();
  });
  it("Divide operation", () => {
    store.dispatch(resetData());
    store.dispatch(addDigit("490"));
    store.dispatch(setOperation("/"));
    store.dispatch(addDigit("7"));
    renderWithProviders(<DisplayCalculator />, { store });
    fireEvent.click(screen.getByRole("button", { name: "=" }));
    expect(screen.getByText("70")).toBeInTheDocument();
  });
  it("Divide by zero", () => {
    store.dispatch(resetData());
    store.dispatch(addDigit("490"));
    store.dispatch(setOperation("/"));
    store.dispatch(addDigit("0"));
    renderWithProviders(<DisplayCalculator />, { store });
    fireEvent.click(screen.getByRole("button", { name: "=" }));
    expect(screen.getByText("Не определено")).toBeInTheDocument();
  });
});

describe("Test Input", () => {
  it("Press number 1234567890", () => {
    store.dispatch(resetData());
    renderWithProviders(<DisplayCalculator />, { store });
    fireEvent.click(screen.getByRole("button", { name: "1" }));
    fireEvent.click(screen.getByRole("button", { name: "2" }));
    fireEvent.click(screen.getByRole("button", { name: "3" }));
    fireEvent.click(screen.getByRole("button", { name: "4" }));
    fireEvent.click(screen.getByRole("button", { name: "5" }));
    fireEvent.click(screen.getByRole("button", { name: "6" }));
    fireEvent.click(screen.getByRole("button", { name: "7" }));
    fireEvent.click(screen.getByRole("button", { name: "8" }));
    fireEvent.click(screen.getByRole("button", { name: "9" }));
    fireEvent.click(screen.getByRole("button", { name: "0" }));
    expect(screen.getByText("1234567890")).toBeInTheDocument();
  });
});

describe("Manual Operation", () => {
  it("Press 11 + 22", () => {
    store.dispatch(resetData());
    renderWithProviders(<DisplayCalculator />, { store });
    fireEvent.click(screen.getByRole("button", { name: "1" }));
    fireEvent.click(screen.getByRole("button", { name: "1" }));
    fireEvent.click(screen.getByRole("button", { name: "+" }));
    fireEvent.click(screen.getByRole("button", { name: "2" }));
    fireEvent.click(screen.getByRole("button", { name: "2" }));
    fireEvent.click(screen.getByRole("button", { name: "=" }));
    expect(screen.getByText("33")).toBeInTheDocument();
  });
  it("Press 11 + 22 + 33", () => {
    store.dispatch(resetData());
    renderWithProviders(<DisplayCalculator />, { store });
    fireEvent.click(screen.getByRole("button", { name: "1" }));
    fireEvent.click(screen.getByRole("button", { name: "1" }));
    fireEvent.click(screen.getByRole("button", { name: "+" }));
    fireEvent.click(screen.getByRole("button", { name: "2" }));
    fireEvent.click(screen.getByRole("button", { name: "2" }));
    fireEvent.click(screen.getByRole("button", { name: "+" }));
    fireEvent.click(screen.getByRole("button", { name: "3" }));
    fireEvent.click(screen.getByRole("button", { name: "3" }));
    fireEvent.click(screen.getByRole("button", { name: "=" }));
    expect(screen.getByText("66")).toBeInTheDocument();
  });
});
