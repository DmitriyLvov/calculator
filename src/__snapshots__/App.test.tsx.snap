// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Render Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="component"
      >
        <div>
          <div
            class="component"
          >
            <div
              class="component"
              style="width: 72px;"
            >
              <button
                class="component_secondary"
                name="9"
              >
                9
              </button>
            </div>
            <div
              class="component"
              style="width: 72px;"
            >
              <button
                class="component_secondary"
                name="8"
              >
                8
              </button>
            </div>
            <div
              class="component"
              style="width: 72px;"
            >
              <button
                class="component_secondary"
                name="7"
              >
                7
              </button>
            </div>
            <div
              class="component"
              style="width: 72px;"
            >
              <button
                class="component_secondary"
                name="6"
              >
                6
              </button>
            </div>
            <div
              class="component"
              style="width: 72px;"
            >
              <button
                class="component_secondary"
                name="5"
              >
                5
              </button>
            </div>
            <div
              class="component"
              style="width: 72px;"
            >
              <button
                class="component_secondary"
                name="4"
              >
                4
              </button>
            </div>
            <div
              class="component"
              style="width: 72px;"
            >
              <button
                class="component_secondary"
                name="3"
              >
                3
              </button>
            </div>
            <div
              class="component"
              style="width: 72px;"
            >
              <button
                class="component_secondary"
                name="2"
              >
                2
              </button>
            </div>
            <div
              class="component"
              style="width: 72px;"
            >
              <button
                class="component_secondary"
                name="1"
              >
                1
              </button>
            </div>
            <div
              class="component"
              style="width: 72px;"
            >
              <button
                class="component_secondary"
                name=","
              >
                ,
              </button>
            </div>
            <div
              class="component"
              style="width: 152px;"
            >
              <button
                class="component_secondary"
                name="0"
              >
                0
              </button>
            </div>
          </div>
        </div>
        <div>
          <div
            class="component"
          >
            <button
              class="component_primary"
              name="="
            >
              =
            </button>
          </div>
        </div>
        <div>
          <div
            class="component"
          >
            <div
              class="component"
            >
              <button
                class="component_secondary"
                name="/"
              >
                /
              </button>
            </div>
            <div
              class="component"
            >
              <button
                class="component_secondary"
                name="x"
              >
                x
              </button>
            </div>
            <div
              class="component"
            >
              <button
                class="component_secondary"
                name="-"
              >
                -
              </button>
            </div>
            <div
              class="component"
            >
              <button
                class="component_secondary"
                name="+"
              >
                +
              </button>
            </div>
          </div>
        </div>
        <div>
          <div
            class="component"
          >
            <p
              class="display_large"
            >
              0
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="component"
    >
      <div>
        <div
          class="component"
        >
          <div
            class="component"
            style="width: 72px;"
          >
            <button
              class="component_secondary"
              name="9"
            >
              9
            </button>
          </div>
          <div
            class="component"
            style="width: 72px;"
          >
            <button
              class="component_secondary"
              name="8"
            >
              8
            </button>
          </div>
          <div
            class="component"
            style="width: 72px;"
          >
            <button
              class="component_secondary"
              name="7"
            >
              7
            </button>
          </div>
          <div
            class="component"
            style="width: 72px;"
          >
            <button
              class="component_secondary"
              name="6"
            >
              6
            </button>
          </div>
          <div
            class="component"
            style="width: 72px;"
          >
            <button
              class="component_secondary"
              name="5"
            >
              5
            </button>
          </div>
          <div
            class="component"
            style="width: 72px;"
          >
            <button
              class="component_secondary"
              name="4"
            >
              4
            </button>
          </div>
          <div
            class="component"
            style="width: 72px;"
          >
            <button
              class="component_secondary"
              name="3"
            >
              3
            </button>
          </div>
          <div
            class="component"
            style="width: 72px;"
          >
            <button
              class="component_secondary"
              name="2"
            >
              2
            </button>
          </div>
          <div
            class="component"
            style="width: 72px;"
          >
            <button
              class="component_secondary"
              name="1"
            >
              1
            </button>
          </div>
          <div
            class="component"
            style="width: 72px;"
          >
            <button
              class="component_secondary"
              name=","
            >
              ,
            </button>
          </div>
          <div
            class="component"
            style="width: 152px;"
          >
            <button
              class="component_secondary"
              name="0"
            >
              0
            </button>
          </div>
        </div>
      </div>
      <div>
        <div
          class="component"
        >
          <button
            class="component_primary"
            name="="
          >
            =
          </button>
        </div>
      </div>
      <div>
        <div
          class="component"
        >
          <div
            class="component"
          >
            <button
              class="component_secondary"
              name="/"
            >
              /
            </button>
          </div>
          <div
            class="component"
          >
            <button
              class="component_secondary"
              name="x"
            >
              x
            </button>
          </div>
          <div
            class="component"
          >
            <button
              class="component_secondary"
              name="-"
            >
              -
            </button>
          </div>
          <div
            class="component"
          >
            <button
              class="component_secondary"
              name="+"
            >
              +
            </button>
          </div>
        </div>
      </div>
      <div>
        <div
          class="component"
        >
          <p
            class="display_large"
          >
            0
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
